sstables/mx/reader.cc

```diff
@@ -1808,5 +1809,262 @@ void mp_row_consumer_reader_mx::on_next_partition(dht::decorated_key key, tombst
+        case bound_kind_m::excl_end_incl_start:
+            return consume_range_tombstone(ecp, bound_kind::incl_start, start_tombstone);
+        default:
+            assert(false && "Invalid boundary type");
+        }
+    }
+
+    data_consumer::proceed consume_row_end() {
+        sstlog.trace("validating_consumer {}: {}()", fmt::ptr(this), __FUNCTION__);
+        if (_expected_clustering_block) {
+            return data_consumer::proceed(!_expected_clustering_block->done);
+        } else {
+            return data_consumer::proceed(!need_preempt());
+        }
+    }
+
+    void on_end_of_stream() {
+        if (!_validator.on_end_of_stream()) {
+            ++_error_count;
+            _error_handler(invalid_end_of_stream{_validator});
+        }
+        sstlog.trace("validating_consumer {}: {}()", fmt::ptr(this), __FUNCTION__);
+    }
+
+    data_consumer::proceed consume_partition_end() {
+        sstlog.trace("validating_consumer {}: {}()", fmt::ptr(this), __FUNCTION__);
+        _current_pos = position_in_partition(position_in_partition::end_of_partition_tag_t{});
+        validate_fragment_order(mutation_fragment_v2::kind::partition_end, {});
+        return data_consumer::proceed::no;
+    }
+
+    void report_unrecoverable_error(std::exception_ptr ex) {
+        ++_error_count;
+        _error_handler(unrecoverable_error{format("{}", ex)});
+    }
+
+    void report_index_data_position_mismach_error(uint64_t index_pos, uint64_t data_pos) {
+        ++_error_count;
+        _error_handler(index_data_position_mismatch_error{index_pos, data_pos});
+    }
+};
+
+future<uint64_t> validate(
+        shared_sstable sstable,
+        schema_ptr schema,
+        reader_permit permit,
+        const io_priority_class& pc,
+        abort_source& abort,
+        validation_error_handler_fn error_handler) {
+
+    validating_consumer consumer(schema, permit, pc, sstable, std::move(error_handler));
+    auto context = data_consume_rows<data_consume_rows_context_m<validating_consumer>>(*schema, sstable, consumer);
+
+    sstables::index_reader idx_reader(sstable, permit, pc, {}, sstables::use_caching::no);
+
+    try {
+        uint64_t current_partition_pos = 0;
+
+        while (!context->eof()) {
+            co_await idx_reader.read_partition_data();
+
+            auto* idx_cursor = idx_reader.current_clustered_cursor();
+
+            const auto index_pos = idx_reader.get_data_file_position();
+            const auto data_pos = context->position();
+            sstlog.trace("validate(): index-data position check for partition {}: {} == {}", idx_reader.get_partition_key(), data_pos, index_pos);
+            if (index_pos != data_pos) {
+                consumer.report_index_data_position_mismach_error(index_pos, data_pos);
+            }
+            current_partition_pos = data_pos;
+
+            consumer.set_expected_partition(idx_reader.get_partition_key());
```

Comment:
Let's mark the consumer's index-related functions with "index" so it's easy to keep track. We're logically merging two streams here.

Reply:
Index key monotonicity is already validated by way of requiring the sstable key order to be that of the index and we validate the sstable key order.